### PR TITLE
TensorProductMatrixSymmetricSum for matrices with rows/columns filled with zero

### PR DESCRIPTION
**Pull request by peterrum**
In my case, these rows/columns correspond to constrained DoFs.

~I still need to add a test.~

### CODE REVIEW

**Comment by peterrum** [comment]
This PR is ready from my side.

File: include/deal.II/lac/tensor_product_matrix.h [hunk 1]
```diff
@@ -450,13 +451,32 @@ namespace internal
...
+
+              if (mm == nn && in[c] == 0.0)
+                {
+                  out[mm + nn * m]     = 1.0;
+                  constrained_dofs[mm] = true;
+                }
+              else
+                out[mm + nn * m] = in[c];
+            }
       };
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:480]: I do not really understand this function. If I read the code correctly, you assume that if the diagonal of the given matrix (either mass matrix or derivative matrix), then the whole row and column need to be zero, so setting the diagonal to 1 is fine. Is this correct? I find this dangerous, even though not mathematically wrong: For a positive definite matrix, which is what we assume here, each diagonal entry needs to be positive. As a result, we could only get a zero if the whole row and column would indeed be zero.

For doing this, I think the code is ugly. Why not simply looping through the diagonal of the mass matrix first and record the constrained entries outside the lambda function, and then use the lambda function that does the copy and sets the diagonal to 1 in both matrices?

File: include/deal.II/lac/tensor_product_matrix.h [hunk 2]
```diff
@@ -450,13 +451,37 @@ namespace internal
...
+                  Assert(mass_matrix[j + i * n_rows] == 0, ExcInternalError());
+                  Assert(derivative_matrix[i + j * n_rows] == 0,
+                         ExcInternalError());
+                  Assert(derivative_matrix[j + i * n_rows] == 0,
+                         ExcInternalError());
+                }
+
+              constrained_dofs[i] = true;
+            }
+        }
```
Comment (peterrum) [inline include/deal.II/lac/tensor_product_matrix.h:475]: @kronbichler I have refactored the setup of the constrained mask. What do you think?

File: include/deal.II/lac/tensor_product_matrix.h [hunk 3]
```diff
@@ -450,13 +451,37 @@ namespace internal
...
-          for (unsigned int nn = 0; nn < n; ++nn)
-            out[mm + nn * m] = *(in++);
+      std::vector<bool> constrained_dofs(n_rows, false);
+
+      for (unsigned int i = 0; i < n_rows; ++i)
+        {
+          if (mass_matrix[i + i * n_rows] == 0.0)
+            {
+              Assert(derivative_matrix[i + i * n_rows] == 0.0,
+                     ExcInternalError());
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:461]: You can skip this check, because it is included in the check below.

File: include/deal.II/lac/tensor_product_matrix.h [hunk 4]
```diff
@@ -450,13 +451,37 @@ namespace internal
...
+        }
+
+      const auto transpose_fill_nm = [&constrained_dofs](Number *           out,
+                                                         const Number *     in,
+                                                         const unsigned int n,
+                                                         const unsigned int m) {
+        for (unsigned int mm = 0, c = 0; mm < m; ++mm)
+          for (unsigned int nn = 0; nn < n; ++nn, ++c)
+            out[mm + nn * m] =
+              (mm == nn && constrained_dofs[mm]) ? Number(1.0) : in[c];
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:484]: Much better.

Comment (kronbichler) [reply include/deal.II/lac/tensor_product_matrix.h:484]: I also prefer the new variable `c` rather than pointer arithmetic. You could also have written `in[mm * n + nn]` in my opinion, but this is up to personal preference.

File: tests/lac/tensor_product_matrix_07.cc [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+void
+do_test(const bool zero_out_constraints)
+{
+  using Number                 = double;
+  const unsigned int dim       = 2;
+  const unsigned int n_rows_1d = 4;
+
+  // compute 2D stiffness matrix
+  const auto refernce_matrices_2D =
```
Comment (kronbichler) [inline tests/lac/tensor_product_matrix_07.cc:96]:
```suggestion
  const auto reference_matrices_2D =
```

File: tests/lac/tensor_product_matrix_07.cc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    compute_reference_matrices<dim, Number>(n_rows_1d - 1);
+  auto K_2D = std::get<1>(refernce_matrices_2D);
+
+  // ... and apply DBC
+  for (unsigned int j = 0; j < Utilities::pow(n_rows_1d, dim); ++j)
+    for (unsigned int i = 0; i < Utilities::pow(n_rows_1d, dim - 1); ++i)
```
Comment (kronbichler) [inline tests/lac/tensor_product_matrix_07.cc:102]: I believe the reason to set two different loop lengths here is that you pick boundary conditions on a single surface in the lexicographic order (x-direction), and leave the rest unchanged. This is simple to miss, so I would like to have a more extensive comment or an descriptive variable name of the loop length referring the unknowns at the identified boundary.

File: tests/lac/tensor_product_matrix_07.cc [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  QGauss<dim>    quadrature(fe_degree + 1);
+  FE_DGQ<dim>    fe(fe_degree);
+
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(fe);
+
+  const unsigned int n_dofs_1D = fe.n_dofs_per_cell();
```
Comment (kronbichler) [inline tests/lac/tensor_product_matrix_07.cc:49]: Why do you have the suffix `1D` when this is a `dim`-dimensional `FE_DGQ`?

File: tests/lac/tensor_product_matrix_07.cc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+      src[i] = 1.0;
+
+      fdm.vmult(make_array_view(dst), make_array_view(src));
+
+      for (unsigned int j = 0; j < fdm.m(); ++j)
+        matrix[j][i] = dst[j];
+    }
+
+  print(matrix);
```
Comment (kronbichler) [inline tests/lac/tensor_product_matrix_07.cc:179]: I think printing some text describing what the respective matrix in the output stream refers to would be helpful.

File: tests/lac/tensor_product_matrix_07.cc [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+
+  return {mass_matrix_reference, derivative_matrix_reference};
+}
+
+template <typename Number>
+void
+print(const FullMatrix<Number> &matrix)
+{
+  matrix.print_formatted(deallog.get_file_stream(), 2, false, 7);
```
Comment (kronbichler) [inline tests/lac/tensor_product_matrix_07.cc:83]: Is any of the numbers you print sensitive to roundoff, say is any of the `0.13` printed below actually `0.125`? That would make the test sensitive to roundoff.

**Review by kronbichler — APPROVED** [review]

File: tests/lac/tensor_product_matrix_07.cc (hunk 9 above)
Comment (peterrum) [reply tests/lac/tensor_product_matrix_07.cc]: I have switched to scientific notation and increased the precision.

File: tests/lac/tensor_product_matrix_07.cc (hunk 6 above)
Comment (peterrum) [reply tests/lac/tensor_product_matrix_07.cc:105]: I have added a comment.
